ledger/catchpointwriter.go

```diff
@@ -0,0 +1,292 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"database/sql"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/algorand/go-codec/codec"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+const (
+	// BalancesPerCatchpointFileChunk defines the number of accounts that would be stored in each chunk in the catchpoint file.
+	// note that the last chunk would typically be less than this number.
+	BalancesPerCatchpointFileChunk = 512
+
+	// catchpointFileVersion is the catchpoint file version
+	catchpointFileVersion = uint64(0200)
+
+	// MaxEncodedAccountDataSize is a rough estimate for the worst-case scenario we're going to have of the account data and address serialized.
+	MaxEncodedAccountDataSize = 64 * 1024
```

Comment:
This is too small. An account with 1,000 created assets could use, just for asset params (looking at the `AssetParams` struct), `1000 * (8 + 4 + 1 + 8 + 32 + 32 + 32 + 32 + 32 + 32 + 32) == 245,000` bytes. I think 500k would be a safe number for today's system.

Reply:
I will create a unit test to measure this number.

Reply:
The actual number is surprisingly high : 324205; I have added a unit test to verify that.